app/models/concerns/string_validation.rb

```diff
@@ -0,0 +1,37 @@
+module StringValidation
+  extend ActiveSupport::Concern
+
+  # Take an object type and a title and return a key that can be used to
+  # store the object in the details hash.
+  #
+  # If the title is already present in the details hash, a counter will be
+  # appended to the key.
+  def key_for_object(object_type, title)
+    base_key = convert_to_valid_key(title, object_type)
+    key = base_key
+    counter = 1
+
+    while details.dig(object_type, key).present?
+      key = "#{base_key}-#{counter}"
+      counter += 1
+    end
+
+    key
+  end
+
+  def convert_to_valid_key(str, fallback)
```

Comment:
is "valid" perhaps a confusing term to use here in `convert_to_valid_key`? Would something like`sanitised_key` be clearer?